python/ray/serve/tests/test_deploy_app.py

```diff
@@ -0,0 +1,1285 @@
+import subprocess
+import sys
+import time
+from contextlib import contextmanager
+from typing import Dict
+from functools import partial
+
+import pytest
+import requests
+
+import ray
+import ray.actor
+import ray._private.state
+from ray.util.state import list_actors, list_tasks
+
+from ray import serve
+from ray._private.test_utils import (
+    wait_for_condition,
+    SignalActor,
+)
+from ray.serve.exceptions import RayServeException
+from ray.serve._private.client import ServeControllerClient
+from ray.serve._private.common import ApplicationStatus, DeploymentStatus, ReplicaState
+from ray.serve._private.constants import (
+    SERVE_NAMESPACE,
+    SERVE_DEFAULT_APP_NAME,
+    DEPLOYMENT_NAME_PREFIX_SEPARATOR,
+)
+from ray.serve.schema import (
+    ServeApplicationSchema,
+    ServeDeploySchema,
+    ServeInstanceDetails,
+)
+from ray.tests.conftest import call_ray_stop_only  # noqa: F401
+
+
+@pytest.fixture
+def shutdown_ray_and_serve():
+    serve.shutdown()
+    if ray.is_initialized():
+        ray.shutdown()
+    yield
+    serve.shutdown()
+    if ray.is_initialized():
+        ray.shutdown()
+
+
+@contextmanager
+def start_and_shutdown_ray_cli():
+    subprocess.check_output(["ray", "stop", "--force"])
+    wait_for_condition(_check_ray_stop, timeout=15)
+    subprocess.check_output(["ray", "start", "--head"])
+
+    yield
+
+    subprocess.check_output(["ray", "stop", "--force"])
+    wait_for_condition(_check_ray_stop, timeout=15)
+
+
+@pytest.fixture(scope="class")
+def start_and_shutdown_ray_cli_class():
+    with start_and_shutdown_ray_cli():
+        yield
+
+
+def _check_ray_stop():
+    try:
+        requests.get("http://localhost:52365/api/ray/version")
+        return False
+    except Exception:
+        return True
+
+
+class TestDeployApp:
```

Comment:
@edoakes sorry I should have noted. This file is mostly just splitting this test class into it's own file.